File: link_checker/robots.py
"""robots.txt parser and compliance checker.

Implements robots.txt protocol for professional crawler etiquette.
Prevents getting IP banned by respecting site policies.
"""

import logging
import time
from dataclasses import dataclass, field
from typing import Optional
from urllib.parse import urlparse

import requests

logger = logging.getLogger(__name__)


@dataclass
class RobotsRule:
    """A single robots.txt rule."""
    user_agent: str
    allow_paths: list[str] = field(default_factory=list)
    disallow_paths: list[str] = field(default_factory=list)
    crawl_delay: Optional[float] = None


class RobotsTxtParser:
    """Parses robots.txt and checks URL permissions."""

    def __init__(
        self,
        user_agent: str = 'LinkCanary',
        timeout: int = 10,
    ):
        self.user_agent = user_agent.lower()
        self.timeout = timeout
        self._cache: dict[str, list[RobotsRule]] = {}
        self._fetched_domains: set[str] = set()

    def fetch_robots_txt(self, base_url: str) -> Optional[str]:
        """
        Fetch robots.txt for a domain.

        Args:
            base_url: Base URL to fetch robots.txt from

        Returns:
            robots.txt content or None if not found
        """
        parsed = urlparse(base_url)
        robots_url = f"{parsed.scheme}://{parsed.netloc}/robots.txt"

        try:
            response = requests.get(
                robots_url,
                timeout=self.timeout,
                headers={'User-Agent': self.user_agent},
            )

            if response.status_code == 200:
                logger.info(f"Found robots.txt at {robots_url}")
                return response.text
            else:
                logger.debug(f"No robots.txt at {robots_url} (status: {response.status_code})")
                return None

        except requests.RequestException as e:
            logger.debug(f"Failed to fetch robots.txt: {e}")
            return None

    def parse_robots_txt(self, content: str) -> list[RobotsRule]:
        """
        Parse robots.txt content into rules.

        Args:
            content: robots.txt text content

        Returns:
            List of RobotsRule objects
        """
        rules = []
        current_rule = None

        for line in content.splitlines():
            line = line.strip()

            # Skip comments and empty lines
            if not line or line.startswith('#'):
                continue

            # Parse directive
            if ':' not in line:
                continue

            directive, value = line.split(':', 1)
            directive = directive.strip().lower()
            value = value.strip()

            if directive == 'user-agent':
                # Start new rule group
                if current_rule:
                    rules.append(current_rule)
                current_rule = RobotsRule(user_agent=value.lower())

            elif current_rule:
                if directive == 'allow':
                    current_rule.allow_paths.append(value)
                elif directive == 'disallow':
                    current_rule.disallow_paths.append(value)
                elif directive == 'crawl-delay':
                    try:
                        current_rule.crawl_delay = float(value)
                    except ValueError:
                        pass

        # Add final rule
        if current_rule:
            rules.append(current_rule)

        return rules

    def get_rules_for_domain(self, base_url: str) -> list[RobotsRule]:
        """
        Get robots.txt rules for a domain (cached).

        Args:
            base_url: Base URL to get rules for

        Returns:
            List of applicable RobotsRule objects
        """
        parsed = urlparse(base_url)
        domain = f"{parsed.scheme}://{parsed.netloc}"

        if domain not in self._cache:
            content = self.fetch_robots_txt(base_url)
            if content:
                self._cache[domain] = self.parse_robots_txt(content)
            else:
                self._cache[domain] = []
            self._fetched_domains.add(domain)

        return self._cache.get(domain, [])

    def _matches_pattern(self, path: str, pattern: str) -> bool:
        """
        Check if a path matches a robots.txt pattern.

        Supports:
        - * matches any sequence of characters
        - $ matches end of URL

        Args:
            path: URL path to check
            pattern: robots.txt pattern

        Returns:
            True if path matches pattern
        """
        if pattern == '/':
            return True

        # Handle wildcards
        if '*' in pattern or '$' in pattern:
            # Convert robots.txt pattern to regex-like matching
            import re
            # Escape special regex chars except * and $
            regex_pattern = ''
            for char in pattern:
                if char == '*':
                    regex_pattern += '.*'
                elif char == '$':
                    regex_pattern += '$'
                elif char in '.^+?{}[]|()\\':
                    regex_pattern += '\\' + char
                else:
                    regex_pattern += char

            if not regex_pattern.endswith('$'):
                regex_pattern += '.*'

            return bool(re.match(regex_pattern, path))

        # Simple prefix match
        return path.startswith(pattern)

    def is_allowed(self, url: str, base_url: str) -> tuple[bool, str]:
        """
        Check if a URL is allowed by robots.txt.

        Args:
            url: URL to check
            base_url: Base URL for the site

        Returns:
            Tuple of (is_allowed, reason)
        """
        rules = self.get_rules_for_domain(base_url)

        if not rules:
            return True, "no robots.txt"

        parsed = urlparse(url)
        path = parsed.path or '/'
        if parsed.query:
            path += '?' + parsed.query

        # Find applicable rules (most specific user-agent match)
        applicable_rules = []
        for rule in rules:
            # Check if rule applies to our user-agent
            if rule.user_agent == '*':
                applicable_rules.append(rule)
            elif self.user_agent in rule.user_agent or rule.user_agent in self.user_agent:
                applicable_rules.insert(0, rule)  # More specific rules first

        # If no applicable rules, allowed
        if not applicable_rules:
            return True, "no applicable rules"

        # Check rules in order (most specific first)
        for rule in applicable_rules:
            # Check Allow directives first (they take precedence)
            for allow_pattern in rule.allow_paths:
                if self._matches_pattern(path, allow_pattern):
                    return True, f"allowed by pattern: {allow_pattern}"

            # Check Disallow directives
            for disallow_pattern in rule.disallow_paths:
                if disallow_pattern and self._matches_pattern(path, disallow_pattern):
                    return False, f"disallowed by pattern: {disallow_pattern}"

        return True, "no matching disallow rules"

    def get_crawl_delay(self, base_url: str) -> Optional[float]:
        """
        Get crawl delay from robots.txt.

        Args:
            base_url: Base URL for the site

        Returns:
            Crawl delay in seconds or None
        """
        rules = self.get_rules_for_domain(base_url)

        for rule in rules:
            if rule.user_agent == '*' or self.user_agent in rule.user_agent:
                if rule.crawl_delay is not None:
                    return rule.crawl_delay

        return None

    def get_stats(self) -> dict:
        """Get robots.txt statistics."""
        return {
            'domains_fetched': len(self._fetched_domains),
            'cached_domains': list(self._fetched_domains),
        }


class RobotsComplianceChecker:
    """High-level robots.txt compliance checker for crawlers."""

    def __init__(
        self,
        user_agent: str = 'LinkCanary/1.0',
        timeout: int = 10,
        ignore_robots: bool = False,
    ):
        self.parser = RobotsTxtParser(user_agent=user_agent, timeout=timeout)
        self.ignore_robots = ignore_robots
        self._skipped_urls: dict[str, str] = {}  # url -> reason

    def check_url(self, url: str, base_url: str) -> tuple[bool, str]:
        """
        Check if a URL should be crawled.

        Args:
            url: URL to check
            base_url: Base URL for the site

        Returns:
            Tuple of (should_crawl, reason)
        """
        if self.ignore_robots:
            return True, "robots.txt ignored"

        is_allowed, reason = self.parser.is_allowed(url, base_url)

        if not is_allowed:
            self._skipped_urls[url] = reason
            logger.info(f"Skipping {url}: {reason}")

        return is_allowed, reason

    def filter_urls(self, urls: list[str], base_url: str) -> tuple[list[str], dict[str, str]]:
        """
        Filter URLs based on robots.txt rules.

        Args:
            urls: List of URLs to filter
            base_url: Base URL for the site

        Returns:
            Tuple of (allowed_urls, skipped_urls_with_reasons)
        """
        if self.ignore_robots:
            return urls, {}

        allowed = []
        for url in urls:
            is_allowed, reason = self.check_url(url, base_url)
            if is_allowed:
                allowed.append(url)

        return allowed, self._skipped_urls

    def get_crawl_delay(self, base_url: str) -> Optional[float]:
        """Get recommended crawl delay for a domain."""
        return self.parser.get_crawl_delay(base_url)

    def get_skipped_urls(self) -> dict[str, str]:
        """Get URLs skipped due to robots.txt."""
        return self._skipped_urls.copy()

    def get_stats(self) -> dict:
        """Get compliance statistics."""
        return {
            'ignored': self.ignore_robots,
            'urls_skipped': len(self._skipped_urls),
            **self.parser.get_stats(),
        }

File: crawl-engine/robots.go
package main

import (
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"
)

// --- robots.txt parsing and compliance ---
// Port of link_checker/robots.py

// RobotsRule represents a single robots.txt rule group.
type RobotsRule struct {
	UserAgent    string
	AllowPaths   []string
	DisallowPaths []string
	CrawlDelay   float64 // seconds, -1 = not set
}

// RobotsChecker fetches, parses, and enforces robots.txt rules.
type RobotsChecker struct {
	Client        *http.Client
	UserAgent     string
	IgnoreRobots  bool
	cachedRules   map[string][]RobotsRule // keyed by "scheme://host"
	skippedCount  int
}

func NewRobotsChecker(userAgent string, timeoutSec int, ignore bool) *RobotsChecker {
	return &RobotsChecker{
		Client:       &http.Client{Timeout: time.Duration(timeoutSec) * time.Second},
		UserAgent:    strings.ToLower(userAgent),
		IgnoreRobots: ignore,
		cachedRules:  make(map[string][]RobotsRule),
	}
}

// GetRulesForDomain fetches and caches robots.txt rules for a domain.
func (rc *RobotsChecker) GetRulesForDomain(baseURL string) []RobotsRule {
	if rc.IgnoreRobots {
		return nil
	}
	parsed, err := url.Parse(baseURL)
	if err != nil {
		return nil
	}
	domain := parsed.Scheme + "://" + parsed.Host
	if rules, ok := rc.cachedRules[domain]; ok {
		return rules
	}
	content := rc.fetchRobotsTxt(domain)
	rules := rc.parseRobotsTxt(content)
	rc.cachedRules[domain] = rules
	return rules
}

func (rc *RobotsChecker) fetchRobotsTxt(domain string) string {
	robotsURL := domain + "/robots.txt"
	req, err := http.NewRequest("GET", robotsURL, nil)
	if err != nil {
		return ""
	}
	req.Header.Set("User-Agent", rc.UserAgent)
	resp, err := rc.Client.Do(req)
	if err != nil {
		return ""
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return ""
	}
	buf := make([]byte, 0, 4096)
	tmp := make([]byte, 4096)
	for {
		n, err := resp.Body.Read(tmp)
		if n > 0 {
			buf = append(buf, tmp[:n]...)
		}
		if err != nil {
			break
		}
	}
	return string(buf)
}

func (rc *RobotsChecker) parseRobotsTxt(content string) []RobotsRule {
	var rules []RobotsRule
	var current *RobotsRule
	for _, line := range strings.Split(content, "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		idx := strings.Index(line, ":")
		if idx < 0 {
			continue
		}
		directive := strings.ToLower(strings.TrimSpace(line[:idx]))
		value := strings.TrimSpace(line[idx+1:])
		switch directive {
		case "user-agent":
			if current != nil {
				rules = append(rules, *current)
			}
			current = &RobotsRule{UserAgent: strings.ToLower(value), CrawlDelay: -1}
		case "allow":
			if current != nil {
				current.AllowPaths = append(current.AllowPaths, value)
			}
		case "disallow":
			if current != nil {
				current.DisallowPaths = append(current.DisallowPaths, value)
			}
		case "crawl-delay":
			if current != nil {
				var delay float64
				_, err := strings.Fields(value)[0], error(nil)
				if len(strings.Fields(value)) > 0 {
					_, err = regexp.MatchString(`^-?[\d.]+$`, strings.Fields(value)[0])
				}
				if err == nil && len(strings.Fields(value)) > 0 {
					fmt_sscanf(strings.Fields(value)[0], &delay)
					current.CrawlDelay = delay
				}
			}
		}
	}
	if current != nil {
		rules = append(rules, *current)
	}
	return rules
}

// fmt_sscanf is a simple float parser to avoid importing fmt for Sscanf.
func fmt_sscanf(s string, f *float64) {
	// Simple parse: handle common cases
	negative := false
	if strings.HasPrefix(s, "-") {
		negative = true
		s = s[1:]
	}
	var result float64
	decimalSeen := false
	var decimalPlace float64 = 10
	for _, c := range s {
		if c >= '0' && c <= '9' {
			digit := float64(c - '0')
			if !decimalSeen {
				result = result*10 + digit
			} else {
				result += digit / decimalPlace
				decimalPlace *= 10
			}
		} else if c == '.' && !decimalSeen {
			decimalSeen = true
		} else {
			break
		}
	}
	if negative {
		result = -result
	}
	*f = result
}

// matchesPattern checks if a path matches a robots.txt pattern (supports * and $).
func matchesRobotsPattern(path, pattern string) bool {
	if pattern == "/" {
		return true
	}
	if strings.Contains(pattern, "*") || strings.Contains(pattern, "$") {
		var re strings.Builder
		for _, c := range pattern {
			switch c {
			case '*':
				re.WriteString(".*")
			case '$':
				re.WriteString("$")
			default:
				if strings.ContainsRune(".^+?{}[]|()\\", c) {
					re.WriteByte('\\')
				}
				re.WriteRune(c)
			}
		}
		if !strings.HasSuffix(re.String(), "$") {
			re.WriteString(".*")
		}
		r, err := regexp.Compile("^" + re.String())
		if err != nil {
			return false
		}
		return r.MatchString(path)
	}
	return strings.HasPrefix(path, pattern)
}

// IsAllowed checks if a URL is allowed by robots.txt.
func (rc *RobotsChecker) IsAllowed(rawURL, baseURL string) (bool, string) {
	if rc.IgnoreRobots {
		return true, "robots.txt ignored"
	}
	rules := rc.GetRulesForDomain(baseURL)
	if len(rules) == 0 {
		return true, "no robots.txt"
	}
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return true, "parse error"
	}
	path := parsed.Path
	if path == "" {
		path = "/"
	}
	if parsed.RawQuery != "" {
		path += "?" + parsed.RawQuery
	}

	// Find applicable rules (most specific user-agent first)
	var applicable []RobotsRule
	for _, rule := range rules {
		if rule.UserAgent == "*" {
			applicable = append(applicable, rule)
		} else if strings.Contains(rule.UserAgent, rc.UserAgent) || strings.Contains(rc.UserAgent, rule.UserAgent) {
			// Insert at beginning (more specific first)
			applicable = append([]RobotsRule{rule}, applicable...)
		}
	}
	if len(applicable) == 0 {
		return true, "no applicable rules"
	}
	for _, rule := range applicable {
		for _, allowPattern := range rule.AllowPaths {
			if matchesRobotsPattern(path, allowPattern) {
				return true, "allowed by pattern: " + allowPattern
			}
		}
		for _, disallowPattern := range rule.DisallowPaths {
			if disallowPattern != "" && matchesRobotsPattern(path, disallowPattern) {
				return false, "disallowed by pattern: " + disallowPattern
			}
		}
	}
	return true, "no matching disallow rules"
}

// CheckURL returns (shouldCrawl, reason). Increments skippedCount if disallowed.
func (rc *RobotsChecker) CheckURL(rawURL, baseURL string) (bool, string) {
	allowed, reason := rc.IsAllowed(rawURL, baseURL)
	if !allowed {
		rc.skippedCount++
	}
	return allowed, reason
}

// FilterURLs returns (allowed, skipped) URL lists.
func (rc *RobotsChecker) FilterURLs(urls []string, baseURL string) ([]string, map[string]string) {
	if rc.IgnoreRobots {
		return urls, nil
	}
	var allowed []string
	skipped := make(map[string]string)
	for _, u := range urls {
		ok, reason := rc.IsAllowed(u, baseURL)
		if ok {
			allowed = append(allowed, u)
		} else {
			skipped[u] = reason
			rc.skippedCount++
		}
	}
	return allowed, skipped
}

// GetCrawlDelay returns the crawl delay from robots.txt, or -1 if not set.
func (rc *RobotsChecker) GetCrawlDelay(baseURL string) float64 {
	rules := rc.GetRulesForDomain(baseURL)
	for _, rule := range rules {
		if rule.UserAgent == "*" || strings.Contains(rule.UserAgent, rc.UserAgent) {
			if rule.CrawlDelay >= 0 {
				return rule.CrawlDelay
			}
		}
	}
	return -1
}

// GetStats returns robots.txt compliance statistics.
func (rc *RobotsChecker) GetStats() JSONRobotsStats {
	return JSONRobotsStats{
		URLsSkipped: rc.skippedCount,
		Ignored:     rc.IgnoreRobots,
	}
}
